File: 08/nim/part01.nim
# Day 8: Space Image Format
# https://adventofcode.com/2019/day/8

import strutils, sequtils

let
  input = open("./inputs/08.txt")
  width = 25
  height = 6

var
  pixels = toSeq(strip(input.readAll()))
  zeroCount = -1
  selectedOffset = 0

input.close()

var offset = 0
while offset < len(pixels):
  let
    layer = pixels[offset..(offset + (width * height) - 1)]
    localCount = count(layer, '0')

  if zeroCount < 0:
    zeroCount = localCount

  if localCount < zeroCount:
    zeroCount = localCount
    selectedOffset = offset

  offset += width * height

let layer = pixels[selectedOffset..(selectedOffset + (width * height) - 1)]
echo count(layer, '1') * count(layer, '2')

File: 08/ruby/part01.rb
# Day 8: Space Image Format
# https://adventofcode.com/2019/day/8

input = File.read('./inputs/08.txt').chomp.chars

WIDTH  = 25
HEIGHT =  6

zero_count     = nil
selected_layer = nil

while !input.empty?
  layer = input.take(WIDTH * HEIGHT)
  input = input.drop(WIDTH * HEIGHT)
  tmp_count = layer.count("0")
  if zero_count.nil? || tmp_count < zero_count
    zero_count = tmp_count
    selected_layer = layer
  end
end

puts selected_layer.count("1") * selected_layer.count("2")
